---
title: "R Notebook"
output: html_notebook
---

1.2 Estimate the LD correlation matrix

```{r}
GSMRcolnames<-c("SNP","a1","a2","a1_freq","bzx","bzx_se","bzx_pval","bzx_n","bzy","bzy_se","bzy_pval","bzy_n")
eduyears4GSMR<-read.table("/data/neurogen/MRandPD/Results/Input4GSMR/eduyearsGSMRinput", skip = 1, col.names = GSMRcolnames)
write.table(eduyears4GSMR[,c(1,2)], "/data/neurogen/MRandPD/Results/Input4GSMR/eduyears4GSMR_snps.allele", col.names=F, row.names=F, quote=F)
```
1.2 Estimate the LD correlation matrix
#this command was run on linux cluster 
# gcta64 --bfile /data/neurogen/MRandPD/1000GenomsPlinkFormat/allchr --extract eduyears4GSMR_snps.allele --update-ref-allele eduyears4GSMR_snps.allele --recode --out gsmr_eduyears

```{r}
# Estimate LD correlation matrix using R
snp_coeff_id = scan("/data/neurogen/MRandPD/Results/Input4GSMR/gsmr_eduyears.xmat.gz", what="", nlines=1)
snp_coeff = read.table("/data/neurogen/MRandPD/Results/Input4GSMR/gsmr_eduyears.xmat.gz", header=F, skip=2)
```

```{r}
# Match the SNP genotype data with the summary data
snp_id = Reduce(intersect, list(eduyears4GSMR$SNP, snp_coeff_id))
eduyears4GSMR = eduyears4GSMR[match(snp_id, eduyears4GSMR$SNP),]
snp_order = match(snp_id, snp_coeff_id)
snp_coeff_id = snp_coeff_id[snp_order]
snp_coeff = snp_coeff[, snp_order]

# Calculate the LD correlation matrix
ldrho = cor(snp_coeff)

# Check the size of the correlation matrix and double-check if the order of the SNPs in the LD correlation matrix is consistent with that in the GWAS summary data
colnames(ldrho) = rownames(ldrho) = snp_coeff_id

dim(ldrho)
```

2. Standardization
```{r}
snpfreq = eduyears4GSMR$a1_freq             # allele frequencies of the SNPs
bzx = eduyears4GSMR$bzx     # effects of the instruments on risk factor
bzx_se = eduyears4GSMR$bzx_se       # standard errors of bzx
bzx_n = eduyears4GSMR$bzx_n          # GWAS sample size for the risk factor
std_zx = std_effect(snpfreq, bzx, bzx_se, bzx_n)    # perform standardisation
eduyears4GSMR$std_bzx = std_zx$b    # standardized bzx
eduyears4GSMR$std_bzx_se = std_zx$se    # standardized bzx_se
head(eduyears4GSMR)
```
3. GSMR analysis
```{r}
bzx = eduyears4GSMR$std_bzx    # SNP effects on the risk factor
bzx_se = eduyears4GSMR$std_bzx_se    # standard errors of bzx
bzx_pval = eduyears4GSMR$bzx_pval   # p-values for bzx
bzy = eduyears4GSMR$bzy    # SNP effects on the disease
bzy_se = eduyears4GSMR$bzy_se    # standard errors of bzy
bzy_pval = eduyears4GSMR$bzy_pval    # p-values for bzy
n_ref = 16155910    # Sample size of the reference sample
gwas_thresh = 5e-6    # GWAS threshold to select SNPs as the instruments for the GSMR analysis
single_snp_heidi_thresh = 0.01    # p-value threshold for single-SNP-based HEIDI-outlier analysis
multi_snp_heidi_thresh = 0.01    # p-value threshold for multi-SNP-based HEIDI-outlier analysis
nsnps_thresh = 7   # the minimum number of instruments required for the GSMR analysis
heidi_outlier_flag = T    # flag for HEIDI-outlier analysis
ld_r2_thresh = 0.05    # LD r2 threshold to remove SNPs in high LD
ld_fdr_thresh = 0.05   # FDR threshold to remove the chance correlations between the SNP instruments
gsmr2_beta = 0     # 0 - the original HEIDI-outlier method; 1 - the new HEIDI-outlier method that is currently under development 
gsmr_results = gsmr(bzx, bzx_se, bzx_pval, bzy, bzy_se, bzy_pval, ldrho, snp_coeff_id, n_ref, heidi_outlier_flag, gwas_thresh, single_snp_heidi_thresh, multi_snp_heidi_thresh, nsnps_thresh, ld_r2_thresh, ld_fdr_thresh, gsmr2_beta)    # GSMR analysis 
filtered_index=gsmr_results$used_index
cat("The estimated effect of the exposure on outcome: ",gsmr_results$bxy)
cat("P-value for bxy: ", gsmr_results$bxy_pval)
cat("Number of non-significant SNPs: ", length(gsmr_results$weak_snps))
cat("Number of SNPs in high LD ( LD rsq >", ld_r2_thresh, "): ", length(gsmr_results$linkage_snps))
cat("Number of pleiotropic outliers: ", length(gsmr_results$pleio_snps))
```
4. Visualization
```{r}
effect_col = colors()[75]
vals = c(bzx[filtered_index]-bzx_se[filtered_index], bzx[filtered_index]+bzx_se[filtered_index])
xmin = min(vals); xmax = max(vals)
vals = c(bzy[filtered_index]-bzy_se[filtered_index], bzy[filtered_index]+bzy_se[filtered_index])
ymin = min(vals); ymax = max(vals)
par(mar=c(5,5,4,2))
plot(bzx[filtered_index], bzy[filtered_index], pch=20, cex=0.8, bty="n", cex.axis=1.1, cex.lab=1.2,
        col=effect_col, xlim=c(xmin, xmax), ylim=c(ymin, ymax),
        xlab=expression(Years~of~education~(italic(b[zx]))),
        ylab=expression(Parkinsons~disease~(italic(b[zy]))))
abline(a= 0, b= gsmr_results$bxy, lwd=1.5, lty=2, col="dim grey")

nsnps = length(bzx[filtered_index])
for( i in 1:nsnps ) {
    # x axis
    xstart = bzx[filtered_index [i]] - bzx_se[filtered_index[i]]; xend = bzx[filtered_index[i]] + bzx_se[filtered_index[i]]
    ystart = bzy[filtered_index[i]]; yend = bzy[filtered_index[i]]
    segments(xstart, ystart, xend, yend, lwd=1.5, col=effect_col)
    # y axis
    xstart = bzx[filtered_index[i]]; xend = bzx[filtered_index[i]] 
    ystart = bzy[filtered_index[i]] - bzy_se[filtered_index[i]]; yend = bzy[filtered_index[i]] + bzy_se[filtered_index[i]]
    segments(xstart, ystart, xend, yend, lwd=1.5, col=effect_col)
}
```
