---
title: "R Notebook"
output: html_notebook
---

This script is for conducting Harmonization

First, we are going to make sure that all variants that are associated with the trait do so on the same direction (possitive). If not, we are going to flip them. 
```{r}
IVcolnames<-c("RsID","REF","ALT","AF","BETA","SE","PVAL","N")
pathData<-c("/data/neurogen/MRandPD/Results/HarmonizationNallsPDdata/data4Harmony/InstruVar/")
IVNames<-list.files(path = pathData, pattern = NULL)
IVflip <- lapply(paste(pathData,IVNames,sep = "" ), read.table, skip=1, col.names=IVcolnames)
IVNames <- sub("IV","", IVNames)
for(j in 1:length(IVNames)){
trait<-IVflip[[j]]
trait<-tbl_df(trait)
Betapos <- trait[trait$BETA > 0, IVcolnames] 
Betaneg <- trait[trait$BETA < 0, IVcolnames]
Betaneg[,2:3]<- Betaneg[,3:2]
Betaneg[,4]<- 1-Betaneg[,4]
Betaneg[,5]<- Betaneg[,5]*-1
trait<-bind_rows(Betapos,Betaneg)
write.table(trait,paste("/data/neurogen/MRandPD/Results/HarmonizationNallsPDdata/data4Harmony/IVinHarmonyPositive/",IVNames[[j]],sep = ""), sep="\t", quote = FALSE)
}
```

Now, those variants that doesen't have the same Effect Allele coding as their corresponding exposure counterpart. 

```{r}
PDcolnames<-c("CHRPOS","A1","A2","freq","b","se","p","N_cases","N_controls","SNP")
pathData<-c("/data/neurogen/MRandPD/Results/HarmonizationNallsPDdata/data4Harmony/SNPsinPD4Flipping/")
PDSNPsNames<-list.files(path = pathData, pattern = NULL)
PD4flip <- lapply(paste(pathData,PDSNPsNames,sep = "" ), read.table, skip=1, col.names=PDcolnames)
PDSNPsNames <- sub("inPD4flip","", PDSNPsNames)
for(j in 1:length(PDSNPsNames)){
trait<-PD4flip[[j]]
trait<-tbl_df(trait)
trait[,2:3]<- trait[,3:2]
trait[,4]<- 1-trait[,4]
trait[,5]<- trait[,5]*-1
write.table(trait,paste("/data/neurogen/MRandPD/Results/HarmonizationNallsPDdata/data4Harmony/PDSNPsflipped/",PDSNPsNames[[j]],sep = ""),sep="\t", quote = FALSE)
}
```
